### ISSUE TITLE
How to calculate the standard deviation of the recruitment deviations for MCMC

### ISSUE THREAD

**Issue by cgrandin** [pacific-hake/hake-assessment#743]
In `assessment-section.Rnw` in the `Sensitivity analyses` part (roughly line 1385, how do we calculate this in an MCMC context instead of MLE (`sigma_R_info$SD_of_devs`)?

```
The value of $\sigma_r$ was changed from a value of 1.4 (base) to alternative
high (1.6) and low (1.0) states. The low value, $\sigma_r = 1.0$,
resulted in a model where the standard deviation of the MLE estimates of
recruitment deviations in the period with the most informative data was **How to calculate this for MCMC??**
\Sexpr{f(sens.models.1[[which(sens.model.names.1=="Sigma R 1.0")]]$sigma_R_info$SD_of_devs[1], 2)},
suggesting that the data were inconsistent with the lower value of $\sigma_r$.
```

**Comment by aaronmberger-nwfsc** [comment]
So using MLE this value is simply something that is reported when you run SS_output so it is easy to see visually or calculate via the model file using the function you have there.  The path to get these same values in MCMC is a bit different, but represents the same thing.  I don't know how much longer we will be doing these two exact SigmaR sensitivities and thus need the supporting writeup, given we have some research project ideas starting hopefully just after this assessment.  Point being, we may or may not want to automate this calculation if it's just going to change in a year or two (but I leave that up to you).  In short the standard deviation of the recruitment deviations for the Main Period of recruitment deviations (i.e., 1970-2018 or labelled Main_RecrDev_2018 for example) based on the posterior median of each individual main period deviation is 1.50 for the fix_low run and 1.83 for the fix_high run.  

To calculate that I just took the median of the posterior distribution for each Main_REcrDev_YYYY and then took the standard deviation of those medians across the years (1970 to 2018).  Does that make sense?

**Comment by aaronmberger-nwfsc** [comment]
Let me know if you want me to add those values or if you want to create a custom variable.

**Comment by kellijohnson-NOAA** [comment]
I think going forward that these values are good to know in an mcmc or MLE context for models regardless of whether or not you are estimating sigma_r. I will make the code that creates the sigmar_R_info table accessible outside of SS_output so we can generate the values for MCMC but wont add them to list generated by SS_output bc that is huge already. So, then we could just do sapply(sens.models.1, get_sigma_R_info, mcmc = TRUE). To get a table of values by model.

**Comment by cgrandin** [comment]
That sounds great, @kellijohnson-NOAA you should never use `sapply()` because it has no guarantee of the type of object it returns, use `purrr::map_dbl()` instead

**Comment by kellijohnson-NOAA** [comment]
Thanks for the tip @cgrandin (I haven't gotten into purrr yet, but I will eventually). What do you guys think of this, seems like mcmc allows rec devs to vary even more than MLE, no big surprise. BTW the new function will work for any model, MLE or MCMC (in theory).
![image](https://user-images.githubusercontent.com/4108564/106070285-05f39d80-60b9-11eb-8d26-1995d2559db4.png)

**Comment by cgrandin** [comment]
Looks fine to me!
Another tip, when you use any dplyr functions you can just type the column name without quotes and leave out all the `.data["xxx"]]` stuff which makes it harder to read. If it has special characters or a space in the column name use backticks:

```
df <- d %>%
  select(subsets, Var_of_devs, `example(1)`)
```

If you want to move into the tidyverse, replace `lapply` with `map`:

`lapply(list, function)` is equivalent to `map(list, function)`

If you want me to shut up now I can do that lol

**Comment by aaronmberger-nwfsc** [comment]
It would be interesting to see if it matches my "hand" calculations.

**Comment by kellijohnson-NOAA** [comment]
Closed with d81c20dd54435942f0c2f9a251cfc35f4e589872
